Sources/SPMBuildCore/PluginScriptRunner.swift

```diff
@@ -0,0 +1,115 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021-2022 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Basics
+import Foundation
+import PackageModel
+import PackageLoading
+import PackageGraph
+import TSCBasic
+import struct TSCUtility.Triple
+
+/// Implements the mechanics of running and communicating with a plugin (implemented as a set of Swift source files). In most environments this is done by compiling the code to an executable, invoking it as a sandboxed subprocess, and communicating with it using pipes. Specific implementations are free to implement things differently, however.
+public protocol PluginScriptRunner {
+    
+    /// Public protocol function that starts compiling the plugin script to an exectutable. The tools version controls the availability of APIs in PackagePlugin, and should be set to the tools version of the package that defines the plugin (not of the target to which it is being applied). This function returns immediately and then calls the completion handler on the callbackq queue when compilation ends.
+    func compilePluginScript(
+        sources: Sources,
+        toolsVersion: ToolsVersion,
+        observabilityScope: ObservabilityScope
+    ) throws -> PluginCompilationResult
+
+    /// Implements the mechanics of running a plugin script implemented as a set of Swift source files, for use
+    /// by the package graph when it is evaluating package plugins.
+    ///
+    /// The `sources` refer to the Swift source files and are accessible in the provided `fileSystem`. The input is
+    /// a PluginScriptRunnerInput structure.
+    ///
+    /// The text output callback handler will receive free-form output from the script as it's running. Structured
+    /// diagnostics emitted by the plugin will be added to the observability scope.
+    ///
+    /// Every concrete implementation should cache any intermediates as necessary to avoid redundant work.
+    func runPluginScript(
+        sources: Sources,
+        initialMessage: Data,
+        toolsVersion: ToolsVersion,
+        workingDirectory: AbsolutePath,
+        writableDirectories: [AbsolutePath],
+        readOnlyDirectories: [AbsolutePath],
+        fileSystem: FileSystem,
+        observabilityScope: ObservabilityScope,
+        callbackQueue: DispatchQueue,
+        delegate: PluginScriptRunnerDelegate,
+        completion: @escaping (Result<Int32, Error>) -> Void
```

Comment:
do we need two queue here? one for the completion handler and one for the delegate?

Reply:
I'll need to double-check — I just moved this around, it doesn't change functionally from before.

Reply:
Actually, it's the same queue.  It's used for both the delegate and the completion block, and it was the name used for this before this PR as well. Possibly it should be renamed, but a name like `delegateQueue` might suggest it is only used for the delegate.

So a typical client would pass in a serial queue and thus get a sequence of zero or more non-overlapping delegate calls and then finally the completion handler on the same queue.

So there should definitely be only one queue (IMO) but the question is just what the right name is when there is both a delegate and a completion handler.  Callback seemed to be a good generic term for that.